src/framerate_gui.cpp

```diff
@@ -0,0 +1,752 @@
+*/
+
+/** @file framerate_gui.cpp GUI for displaying framerate/game speed information. */
+
+#include "framerate_type.h"
+#include <chrono>
+#include "gfx_func.h"
+#include "window_gui.h"
+#include "table/sprites.h"
+#include "strings_func.h"
+#include "debug.h"
+#include "console_func.h"
+#include "console_type.h"
+
+
+namespace {
+
+	/** Number of data points to keep in buffer for each performance measurement */
+	const int NUM_FRAMERATE_POINTS = 512;
+	/** Units a second is divided into in performance measurements */
+	const int TIMESTAMP_PRECISION = 1000000;
+
+	struct PerformanceData {
+		static const TimingMeasurement INVALID_DURATION = UINT64_MAX;
+
+		TimingMeasurement durations[NUM_FRAMERATE_POINTS];
+		TimingMeasurement timestamps[NUM_FRAMERATE_POINTS];
+		double expected_rate;
+		int next_index;
+		int prev_index;
+		int num_valid;
+
+		explicit PerformanceData(double expected_rate) : expected_rate(expected_rate), next_index(0), prev_index(0), num_valid(0) { }
+
+		void Add(TimingMeasurement start_time, TimingMeasurement end_time)
+		{
+			this->durations[this->next_index] = end_time - start_time;
+			this->timestamps[this->next_index] = start_time;
+			this->prev_index = this->next_index;
+			this->next_index += 1;
+			if (this->next_index >= NUM_FRAMERATE_POINTS) this->next_index = 0;
+			this->num_valid = min(NUM_FRAMERATE_POINTS, this->num_valid + 1);
+		}
+
+		void BeginAccumulate(TimingMeasurement start_time)
+		{
+			this->timestamps[this->next_index] = start_time;
+			this->durations[this->next_index] = 0;
+			this->prev_index = this->next_index;
+			this->next_index += 1;
+			if (this->next_index >= NUM_FRAMERATE_POINTS) this->next_index = 0;
+			this->num_valid = min(NUM_FRAMERATE_POINTS, this->num_valid + 1);
+		}
+
+		void AddAccumulate(TimingMeasurement duration)
+		{
+			this->durations[this->prev_index] += duration;
+		}
+
+		void AddPause(TimingMeasurement start_time)
+		{
+			if (this->durations[this->prev_index] != INVALID_DURATION) {
+				this->timestamps[this->next_index] = start_time;
+				this->durations[this->next_index] = INVALID_DURATION;
+				this->prev_index = this->next_index;
+				this->next_index += 1;
+				if (this->next_index >= NUM_FRAMERATE_POINTS) this->next_index = 0;
+				this->num_valid += 1;
+			}
+		}
+
+		double GetAverageDurationMilliseconds(int count)
+		{
+			count = min(count, this->num_valid);
+
+			int first_point = this->prev_index - count;
+			if (first_point < 0) first_point += NUM_FRAMERATE_POINTS;
+
+			double sumtime = 0;
+			for (int i = first_point; i < first_point + count; i++) {
+				auto d = this->durations[i % NUM_FRAMERATE_POINTS];
+				if (d != INVALID_DURATION) {
+					sumtime += d;
+				} else {
+					count--;
+				}
+			}
+
+			if (count > 0) {
+				return sumtime * 1000 / count / TIMESTAMP_PRECISION;
+			} else {
+				return 0;
+			}
```

Comment:
Can you duff this into:

if (count == 0) return 0;
return sumtime * 1000 / count / TIMESTAMP_PRECISION;